### PR TITLE
MGMT-11293: Added default network type setting during registration

### PR DESCRIPTION
**Pull request by danmanor**
## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by eliorerz** [comment]
Please remove default `network_type` definition from [swagger.yaml](https://github.com/openshift/assisted-service/blob/master/swagger.yaml#L4293-#L4297) (don't forget to run `skipper make generate` after you delete it)

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -340,9 +340,82 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+		// If an OpenShift version is not specified, we do not assume it is >= 4.12
+		if params.NewClusterParams.OpenshiftVersion == nil {
+			isOpenShiftVersionRecentEnough = false
+		} else {
+			openShiftVersion, err := version.NewVersion(swag.StringValue(params.NewClusterParams.OpenshiftVersion))
+			openShiftVersionParam = openShiftVersion
+			if err != nil {
+				return params.NewClusterParams.NetworkType, err
+			}
+			thresholdVersion, err := version.NewVersion("4.12")
```
[the remark this answers is not in the document]
Comment (danmanor) [reply internal/bminventory/inventory.go:372]: Currently version library compares as follow: x.y.z > x.y.z-rc > x.y.z-fc > x.y.z-ec, so it will not catch pre-release versions of 4.12.0. If we wish to do so I can change the threshold version to 4.12.0-ec or use a different function that ignores pre-release data

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -340,9 +340,82 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+	}
+}
+
+// returns whether the IP version is IPv6
+func isIPv6(cidr *string) (bool, error) {
+	// if no ClusterNetworkCidr specified, we do not assume nothing of it
+	if cidr == nil {
+		return false, nil
+	}
+	_, _, err := net.ParseCIDR(*cidr)
```
Comment (gamli75) [inline internal/bminventory/inventory.go:408]: this code is redundant - https://github.com/openshift/assisted-service/blob/a7e4f7485b3b7939e93e7c319865df3a6b18ee68/internal/network/family_utils.go#L16 is doing the same.
you just need to call to network.IsIPV6CIDR

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -340,9 +340,82 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+	if err != nil {
+		return false, errors.Wrapf(err, "%s is not a valid cidr", *cidr)
+	}
+	if network.IsIPV4CIDR(*cidr) {
```
Comment (gamli75) [inline internal/bminventory/inventory.go:412]: the method name is IsIPv6 but you are asking about IPv4(network.IsIPV4CIDR) - it is better to call network.IsIPV6CIDR but as I mentioned above you don't need it

File: internal/bminventory/inventory_test.go [hunk 4]
```diff
@@ -9812,6 +9812,108 @@ var _ = Describe("TestRegisterCluster", func() {
...
+			It("NetworkType default value", func() {
+				type clusterParamsForTest struct {
+					OpenShiftVersion     *string
+					HighAvailabilityMode *string
+					ClusterNetworkCidr   *string
+					DesiredNetworkType   string
+					Platform             models.Platform
+				}
+
+				tests := []clusterParamsForTest{
```
Comment (gamli75) [inline internal/bminventory/inventory_test.go:9828]: please add another test for a case that CIDR is null or empty

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -340,9 +340,82 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+	}
+
 	return params, nil
 }
 
+// If the cluster is SNO, or the IP version is IPv6 (based on ClusterNetworkCidr param), or the OpenShift version >= 4.12,
+// the network type will be set to "OVNKubernetes", otherwise it will be set to "OpenShiftSDN"
+func getDefaultNetworkType(params installer.V2RegisterClusterParams) (*string, error) {
+	// If there is not a network specified already
+	if params.NewClusterParams.NetworkType == nil {
```
Comment (avishayt) [inline internal/bminventory/inventory.go:356]: Avoid nesting by checking "!= nil" and returning early

File: internal/bminventory/inventory.go [hunk 6]
```diff
@@ -340,9 +340,82 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+
+		var (
+			isOpenShiftVersionRecentEnough bool
+			openShiftVersionParam          *version.Version
+		)
+
+		// If an OpenShift version is not specified, we do not assume it is >= 4.12
+		if params.NewClusterParams.OpenshiftVersion == nil {
```
Comment (avishayt) [inline internal/bminventory/inventory.go:364]: OpenshiftVersion should be required.  Anyway, you can also check this up front and return early to avoid more nesting.

File: internal/bminventory/inventory.go [hunk 7]
```diff
@@ -340,9 +341,62 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+	// If one of the conditions match, we will return OVNKubernetes, otherwise OpenShiftSDN
+	if isOpenShiftVersionRecentEnough || isSingleNodeCluster || isIPv6 {
+		return swag.String(models.ClusterCreateParamsNetworkTypeOVNKubernetes), nil
+	} else {
+		return swag.String(models.ClusterCreateParamsNetworkTypeOpenShiftSDN), nil
+	}
+}
+
+func isSingleNodeCluster(params installer.V2RegisterClusterParams) bool {
+	if swag.StringValue(params.NewClusterParams.HighAvailabilityMode) == "None" {
```
Comment (osherdp) [inline internal/bminventory/inventory.go:378]: replace "none" with ``models.ClusterHighAvailabilityModeNone``

File: internal/bminventory/inventory.go [hunk 8]
```diff
@@ -340,9 +341,62 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+		return true
+	} else {
+		return false
+	}
```
Comment (osherdp) [inline internal/bminventory/inventory.go:382]:
```suggestion
	return swag.StringValue(params.NewClusterParams.HighAvailabilityMode) == "None"
```

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -340,9 +341,62 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+
+	isOpenShiftVersionRecentEnough, err := verifyMinimalOpenShiftVersionForDefaultNetworkTypeOVNKubernetes(swag.StringValue(params.NewClusterParams.OpenshiftVersion))
+	if err != nil {
+		return params.NewClusterParams.NetworkType, err
+	}
+
+	isSingleNodeCluster := isSingleNodeCluster(params)
+	isIPv6 := network.IsIPv6CIDR(swag.StringValue(params.NewClusterParams.ClusterNetworkCidr))
+
+	// If one of the conditions match, we will return OVNKubernetes, otherwise OpenShiftSDN
```
Comment (osherdp) [inline internal/bminventory/inventory.go:369]: delete

File: internal/bminventory/inventory.go (hunk 8 above)
Comment (osherdp) [reply internal/bminventory/inventory.go:382]: same goes to other ``return <bool-value>`` cases

**Review by osherdp — CHANGES_REQUESTED** [review]
partial review

File: internal/bminventory/inventory.go [hunk 10]
```diff
@@ -340,9 +341,51 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+		return swag.String(models.ClusterCreateParamsNetworkTypeOpenShiftSDN), nil
+	}
+}
+
+func verifyMinimalOpenShiftVersionForDefaultNetworkTypeOVNKubernetes(requestedOpenshiftVersion string) (bool, error) {
+	ocpVersion, err := version.NewVersion(requestedOpenshiftVersion)
+	if err != nil {
+		return false, errors.Errorf("Failed to parse OCP version %s", requestedOpenshiftVersion)
+	}
+	minimalVersionForSno, err := version.NewVersion(minimalOpenShiftVersionForDefaultNetworkTypeOVNKubernetes)
```
Comment (gamli75) [inline internal/bminventory/inventory.go:379]: why did the member call minimalVersionForSno? it's not relevant to SNO
should be minimalVersionForNetworkTypeOVN or something similar

File: internal/bminventory/inventory.go [hunk 11]
```diff
@@ -340,9 +341,51 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
 
+	params.NewClusterParams.NetworkType, err = getDefaultNetworkType(params)
+	if err != nil {
+		return params, err
+	}
+
 	return params, nil
 }
 
+// If the cluster is SNO, or the IP version is IPv6 (based on ClusterNetworkCidr param), or the OpenShift version >= 4.12,
```
Comment (gamli75) [inline internal/bminventory/inventory.go:352]: in the comment, you wrote that you are checking if "the IP version is IPv6" but I don't see you handle IPv6 in the method. 
I'm not sure what should be the fix: Add check for IPv6 or remove it from the comment

Comment (danmanor) [reply internal/bminventory/inventory.go]: Mistake, was corrected now

File: internal/bminventory/inventory.go (hunk 10 above)
Comment (danmanor) [reply internal/bminventory/inventory.go]: Mistake, was corrected now

File: internal/bminventory/inventory_test.go [hunk 12]
```diff
@@ -10029,6 +10029,160 @@ var _ = Describe("TestRegisterCluster", func() {
...
+					NewClusterParams.Platform = &t.Platform
+
+					reply := bm.V2RegisterCluster(ctx, installer.V2RegisterClusterParams{
+						NewClusterParams: NewClusterParams,
+					})
+
+					Expect(reply).Should(BeAssignableToTypeOf(installer.NewV2RegisterClusterCreated()))
+					actual := reply.(*installer.V2RegisterClusterCreated).Payload
+					Expect(swag.StringValue(actual.NetworkType)).To(Equal(t.DesiredNetworkType))
+				}
```
Comment (osherdp) [inline internal/bminventory/inventory_test.go:10130]: it seems to be the same code as above cases
why not merge ``availability-mode=nil`` to the above dictionary cases?

Comment (osherdp) [reply internal/bminventory/inventory_test.go:10130]: same question for the pre-release cases

Comment (danmanor) [reply internal/bminventory/inventory_test.go:10130]: I wanted to create a semantic separation between the tests, thought it would be too long in one test

File: internal/bminventory/inventory.go [hunk 13]
```diff
@@ -340,9 +341,51 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+	isSingleNodeCluster := swag.StringValue(params.NewClusterParams.HighAvailabilityMode) == models.ClusterCreateParamsHighAvailabilityModeNone
+
+	if isOpenShiftVersionRecentEnough || isSingleNodeCluster {
+		return swag.String(models.ClusterCreateParamsNetworkTypeOVNKubernetes), nil
+	} else {
+		return swag.String(models.ClusterCreateParamsNetworkTypeOpenShiftSDN), nil
+	}
+}
+
+func verifyMinimalOpenShiftVersionForDefaultNetworkTypeOVNKubernetes(requestedOpenshiftVersion string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (danmanor) [reply internal/bminventory/inventory.go:374]: True, but it doesn't make use of Core() ("Core returns a new version constructed from only the MAJOR.MINOR.PATCH segments of the version, without prerelease or metadata.") . we need Core for example in order to determine that 4.12.0-ec.3 >= 4.12. Should I add a new function in internal/common/version.go that uses Core() ?

[the remark this answers is not in the document]
Comment (avishayt) [reply internal/bminventory/inventory.go:374]: The unit tests on that method look convincing:
230:		is410Version, _ := VersionGreaterOrEqual("4.10.0", "4.10.0-0.alpha")
234:		is410Version, _ := VersionGreaterOrEqual("4.10.0-fc.1", "4.10.0-0.alpha")
238:		is410Version, _ := VersionGreaterOrEqual("4.10.1-fc.1", "4.10.0-0.alpha")
242:		is410Version, _ := VersionGreaterOrEqual("4.10.0-0.nightly-2022-01-23-013716", "4.10.0-0.alpha")

If it's broken, then we can fix it, but no need to replicate functionality.

File: swagger.yaml [hunk 14]
```diff
@@ -4294,7 +4294,7 @@ definitions:
         type: string
         description: "The desired network type used."
         enum: ['OpenShiftSDN', 'OVNKubernetes']
-        default: 'OpenShiftSDN'
+        x-nullable: true
```
Comment (avishayt) [inline swagger.yaml:4297]: Does this break clients/API?

Comment (osherdp) [reply swagger.yaml:4297]: I think it shouldn't break
At the schema-level, we don't disallow anything that wasn't possible before
At the data & values level, UI team is preparing to remove their extra-logic of setting ovn on "ocp>=4.12 || SNO", and provide no value for the network-type (letting us decide on cluster creation what should be the smart-default)

File: internal/bminventory/inventory.go [hunk 15]
```diff
@@ -340,9 +341,36 @@ func (b *bareMetalInventory) setDefaultRegisterClusterParams(ctx context.Context
...
+// the network type will be set to "OVNKubernetes", otherwise it will be set to "OpenShiftSDN"
+func getDefaultNetworkType(params installer.V2RegisterClusterParams) (*string, error) {
+	// If the network type is specified already, we return it
+	if params.NewClusterParams.NetworkType != nil {
+		return params.NewClusterParams.NetworkType, nil
+	}
+
+	isOpenShiftVersionRecentEnough, err := common.VersionGreaterOrEqual(swag.StringValue(params.NewClusterParams.OpenshiftVersion), minimalOpenShiftVersionForDefaultNetworkTypeOVNKubernetes)
+	if err != nil {
+		return params.NewClusterParams.NetworkType, err
```
Comment (mkowalski) [inline internal/bminventory/inventory.go:362]: I may be too picky, but I find it slightly confusing. We enter the error-path here and return effectively a tuple (nil,err) but instead of nil we write it as params.somethingsomething. The first argument is effectively nil because if it wasn’t, we would have caught it in the conditional two lines above. So if this is true, could we just write here “return nil, err”? To anyone consuming this function in the future it will be clear that this path is not a good one

File: internal/bminventory/inventory_test.go [hunk 16]
```diff
@@ -10029,6 +10029,111 @@ var _ = Describe("TestRegisterCluster", func() {
 			}
 		}
 
+		// Testing the default network type setting during register flow, with various combinations of platforms,
+		// OpenShift version and high availability modes
+		// Note: clusters with high availability mode "None" must not have have Platform "baremetal", and OpenShift version >= 4.8
```
Comment (mkowalski) [inline internal/bminventory/inventory_test.go:10034]: Where is that sentence coming from? I agree that SNO is only OCP 4.8+, but nothing stops OCP 4.13 to support SNO with platform=baremetal

File: swagger.yaml (hunk 14 above)
Comment (mkowalski) [reply swagger.yaml:4297]: I ran through the code and I also think this is a safe change. If user provided the value, we will just carry it. If user does not provide a value, then immediately when we go to cluster creation we will fill that value with something real.

So in the current form it looks like there is no path to end up with empty network type in cluster creation.

How about cluster update though? I may not read it carefully enough but I have a feeling that it’s now possible to unset network type

File: internal/bminventory/inventory_test.go (hunk 16 above)
Comment (avishayt) [reply internal/bminventory/inventory_test.go:10034]: SNO must be platform "none"

File: swagger.yaml (hunk 14 above)
Comment (danmanor) [reply swagger.yaml:4297]: I don't think it is possible to unset the network type in cluster update, because: 
https://github.com/openshift/assisted-service/blob/2ee5cb8ca37b7cf6d96ec27f4d7980aaeea111e6/internal/bminventory/inventory.go#L2350-L2351

File: internal/bminventory/inventory_test.go [hunk 17]
```diff
@@ -10018,6 +10018,110 @@ var _ = Describe("TestRegisterCluster", func() {
 		ctrl.Finish()
 	})
 
+	// Testing the default network type setting during register flow, with various combinations of platforms,
+	// OpenShift version and high availability modes
+	// Note: SNO currently must have have Platform "none", and OpenShift version >= 4.8
```
Comment (osherdp) [inline internal/bminventory/inventory_test.go:10023]:
```suggestion
	// Note: SNO is supported starting OpenShift v4.8, and currently only supports Platform "none"
```

**Review by osherdp — CHANGES_REQUESTED** [review]

**Review by osherdp — APPROVED** [review]
